lib/frontend.ml

```diff
@@ -144,12 +146,27 @@ module Make(C: S.CONFIGURATION) = struct
   let notify nf () =
     Xen_os.Eventchn.notify h nf.evtchn
 
+  let take_pages t n =
+    let rec inner t n acc =
+      if n==0 then acc
+      else (
+        match t.free_pages with
+        | hd::tl ->
+          t.free_pages <- tl;
+          inner t (n-1) (hd::acc)
+        | [] ->
+          let pages = Io_page.pages n in
+          List.append pages acc
```

Comment:
My first thought was "these new pages are never free'd". But I'm unsure. We always add pages (back) to `free_pages` once we're done using them?

Reply:
Sorry I missread, yes I never free the pages and keep them in case another rush of packets comes un the future.

Reply:
Hmm, it's not obvious to me what to do. If we have a condition I think we could have a deadlock.
Perhaps one way forward could be to pre-allocate N pages (as you do), allow up to N+M pages in the `free_pages` list and if more than N+M pages are allocated then we don't keep them around after they've been used.

I don't quite understand ephemerons and how they work, but they may also be interesting to use.

Reply:
Another thing would be to have some algorithm and threshold for bumping the amount of allocated io-pages for the concrete application (i.e. if we're out of io-page, use the double amount). Again, this is tough since this may open DoS by someone doing whatever (e.g. sending small invalid frames) that leads to lots of io-pages being allocated [and never freed].

Reply:
And an additional note, the reused pages are not blanked anymore, which may leads to information leaks (unsure how usable yet, but..)?